pkg/processor/schemaregistry/avro/schema_test.go

```diff
@@ -0,0 +1,367 @@
+			is.NoErr(err)
+
+			wantSchema := &Schema{
+				schema: must(avro.NewRecordSchema("record", "",
+					[]*avro.Field{must(avro.NewField("foo", tc.wantSchema))},
+				)),
+			}
+			wantSchema.Sort()
+			gotSchema.Sort()
+			is.Equal(wantSchema.String(), gotSchema.String())
+
+			// now try to marshal the value with the schema
+			bytes, err := gotSchema.Marshal(haveValue)
+			is.NoErr(err)
+
+			// unmarshal the bytes back into structured data and compare the value
+			var gotValue record.StructuredData
+			err = gotSchema.Unmarshal(bytes, &gotValue)
+			is.NoErr(err)
+
+			wantValue := newRecord(tc.wantValue)
+			is.Equal(wantValue, gotValue)
+		})
+	}
+}
+
+func TestSchemaForType_NestedStructuredData(t *testing.T) {
+	is := is.New(t)
+
+	have := record.StructuredData{
+		"foo": "bar",
+		"level1": record.StructuredData{
+			"foo": "bar",
+			"level2": record.StructuredData{
+				"foo": "bar",
+				"level3": record.StructuredData{
+					"foo":        "bar",
+					"regularMap": map[string]bool{},
+				},
+			},
+		},
+	}
+
+	want := &Schema{schema: must(avro.NewRecordSchema(
+		"record", "",
+		[]*avro.Field{
+			must(avro.NewField("foo", avro.NewPrimitiveSchema(avro.String, nil))),
+			must(avro.NewField("level1",
+				must(avro.NewRecordSchema(
+					"record.level1", "",
+					[]*avro.Field{
+						must(avro.NewField("foo", avro.NewPrimitiveSchema(avro.String, nil))),
+						must(avro.NewField("level2",
+							must(avro.NewRecordSchema(
+								"record.level1.level2", "",
+								[]*avro.Field{
+									must(avro.NewField("foo", avro.NewPrimitiveSchema(avro.String, nil))),
+									must(avro.NewField("level3",
+										must(avro.NewRecordSchema(
+											"record.level1.level2.level3", "",
+											[]*avro.Field{
+												must(avro.NewField("foo", avro.NewPrimitiveSchema(avro.String, nil))),
+												must(avro.NewField("regularMap", avro.NewMapSchema(
+													avro.NewPrimitiveSchema(avro.Boolean, nil),
+												))),
+											},
+										)),
+									)),
+								},
+							)),
+						)),
+					},
+				)),
+			)),
+		},
+	))}
+	want.Sort()
+
+	got, err := SchemaForType(have)
+	is.NoErr(err)
+	is.Equal(want.String(), got.String())
+
+	bytes, err := got.Marshal(have)
+	is.NoErr(err)
+	// only try to unmarshal to ensure there's no error, other tests assert that
+	// umarshaled data matches the expectations
+	var unmarshaled record.StructuredData
+	err = got.Unmarshal(bytes, &unmarshaled)
+	is.NoErr(err)
+}
+
+func TestSchemaForType_UnsupportedTypes(t *testing.T) {
+	testCases := []any{
+		// avro only supports fixed byte arrays
+		[4]int{},
+		[4]bool{},
+		// avro only supports maps with string keys
+		map[int]string{},
+		map[bool]string{},
+		// avro only supports signed integers
+		uint64(1),
+		uint(1),
+	}
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("%T", tc), func(t *testing.T) {
+			is := is.New(t)
+			_, err := SchemaForType(tc)
+			is.True(err != nil)
```

Comment:
Nitpick: it would be nice if we could check the error itself too (i.e. that it didn't fail for another reason).